2015-Fall/oPolo/Calculator/Calculator/CalculatorBrain.swift

```diff
@@ -0,0 +1,114 @@
+//
+//  CalculatorBrain.swift
+//  Calculator
+//
+//  Created by Jonas Bruun Jacobsen on 15/09/15.
+//  Copyright (c) 2015 Jonas Bruun Jacobsen. All rights reserved.
+//
+
+import Foundation
+
+class CalculatorBrain {
+    private enum Op: Printable {
+        case Operand(Double)
+        case Constant(String, Double)
+        case UnaryOperation(String, Double -> Double)
+        case BinaryOperation(String, (Double, Double) -> Double)
+        
+        var description: String {
+            get {
+                switch self {
+                case .Operand(let operand):
+                    return "\(operand)"
+                case .Constant(let constant, _):
+                    return constant
+                case .UnaryOperation(let symbol, _):
+                    return symbol
+                case .BinaryOperation(let symbol, _):
+                    return symbol
+                }
+            }
+        }
+    }
+    
+    private var opStack = [Op]() {
+        didSet {
+            //Fires a notification when a new operand/operator/constant/etc is added to the stack.
```

Comment:
Surprised me as well. After investigating it a bit, my guess it that since arrays are value types, append results in a new copy of the array. This can be done in `mutating` [functions](https://developer.apple.com/library/prerelease/ios/documentation/Swift/Conceptual/Swift_Programming_Language/Methods.html), like `append`, where `self` is can be set to something new. Still, the semantic seems strange.

You can use the following the see that the memory address changes: 
        
        var arr = [String]()
        print("\(unsafeAddressOf(arr))")
        arr.append("foo")
        print("\(unsafeAddressOf(arr))")
        arr.append("foo2")
        print("\(unsafeAddressOf(arr))")